src/blocks.c

```diff
@@ -485,6 +484,17 @@ static void process_footnotes(cmark_parser *parser) {
         if (!footnote->ix)
           footnote->ix = ++ix;
 
+        // keep track of a) how many times this footnote def has been
+        // referenced, and b) which reference count this footnote ref is at
+        // this is used by renderers when generating links and backreferences.
+        cur->footnote.ix = ++footnote->node->footnote.count;
+
+        // store the footnote reference text label in the footnote ref's node's
+        // `user_data`, so that renderers can use the label when generating
+        // links and backreferences.
+        cur->user_data = parser->mem->calloc(1, (sizeof(char) * cur->as.literal.len) + 1);
```

Comment:
I am not convinced that the memory allocated here is ever freed.

The `S_free_nodes` function will free `user_data`, but only if `user_data_free_func` has been set. I stepped through an execution of `cmark-gfm` in the Xcode debugger (using the footnotes extension, and an input file containing footnotes), and I observed `S_free_nodes` *not* freeing this label, because `user_data_free_func` was `NULL`.

It's possible that setting `user_data_free_func` to be `parser->mem->free` will solve the leak.

However, I'm concerned that the library has public API functions `cmark_node_set_user_data` and `cmark_node_set_user_data_free_func`, suggesting that users of the library are free to overwrite `user_data`, and to specify custom deallocation functions for it. That would mean the text label we just put in `cur->user_data` could possibly be overwritten, or freed with a custom deallocator that doesn't match `calloc`. In that case, we really ought to be using some other, private field for storing the text label.

I'm not at all familiar with this codebase, so it's quite possible I've misunderstood something.

Reply:
On reflection, using `user_data` was a bad move altogether. At the time I felt timid about introducing another member to the `cmark_node` struct, and therefore sought to re-use the existing pointer.

Because of this comment, I went and added a pointer to the parent footnote definition inside the `cmark_node`, and then use that pointer for fetching the footnote label string data as needed during the rendering phase.

I believe the additional pointer should be freed during the normal free step, and then I made sure to immediately free the newly allocated strings just-in-case (tho it's not clear to me that it's strictly necessary, given how potentially allocated memory is treated from what `cmark_chunk_to_cstr` returns elsewhere in the same function).